src/nor_flash.rs

```diff
@@ -0,0 +1,239 @@
+	/// It is not allowed to write to the same word twice.
+	/// `offset` and `bytes.len()` must both be multiples of `write_size()` and properly aligned.
+	fn try_write(&mut self, offset: u32, bytes: &[u8]) -> Result<(), Self::Error>;
+}
+
+/// Marker trait for NorFlash relaxing the restrictions on `write`.
+///
+/// Writes to the same word twice are now allowed. The result is the logical AND of the
+/// previous data and the written data. That is, it is only possible to change 1 bits to 0 bits.
+///
+/// If power is lost during write:
+/// - Bits that were 1 on flash and are written to 1 are guaranteed to stay as 1
+/// - Bits that were 1 on flash and are written to 0 are undefined
+/// - Bits that were 0 on flash are guaranteed to stay as 0
+/// - Rest of the bits in the page are guaranteed to be unchanged
+pub trait MultiwriteNorFlash: NorFlash {}
+
+struct Page {
+	pub start: u32,
+	pub size: usize,
+}
+
+impl Page {
+	fn new(index: u32, size: usize) -> Self {
+		Self {
+			start: index * size as u32,
+			size,
+		}
+	}
+
+	/// The end address of the page
+	const fn end(&self) -> u32 {
+		self.start + self.size as u32
+	}
+}
+
+impl Region for Page {
+	/// Checks if an address offset is contained within the page
+	fn contains(&self, address: u32) -> bool {
+		(self.start <= address) && (self.end() > address)
+	}
+}
+
+///
+pub struct RmwNorFlashStorage<'a, S> {
+	storage: S,
+	merge_buffer: &'a mut [u8],
+}
+
+impl<'a, S> RmwNorFlashStorage<'a, S>
+where
+	S: NorFlash,
+{
+	/// Instantiate a new generic `Storage` from a `NorFlash` peripheral
+	///
+	/// **NOTE** This will panic if the provided merge buffer,
+	/// is smaller than the erase size of the flash peripheral
+	pub fn new(nor_flash: S, merge_buffer: &'a mut [u8]) -> Self {
+		if merge_buffer.len() < S::ERASE_SIZE {
+			panic!("Merge buffer is too small");
+		}
+
+		Self {
+			storage: nor_flash,
+			merge_buffer,
+		}
+	}
+}
+
+impl<'a, S> ReadStorage for RmwNorFlashStorage<'a, S>
+where
+	S: ReadNorFlash,
+{
+	type Error = S::Error;
+
+	fn try_read(&mut self, offset: u32, bytes: &mut [u8]) -> Result<(), Self::Error> {
+		// Nothing special to be done for reads
+		self.storage.try_read(offset, bytes)
+	}
+
+	fn capacity(&self) -> usize {
+		self.storage.capacity()
+	}
+}
+
+impl<'a, S> Storage for RmwNorFlashStorage<'a, S>
+where
+	S: NorFlash,
+{
+	fn try_write(&mut self, offset: u32, bytes: &[u8]) -> Result<(), Self::Error> {
+		// Perform read/modify/write operations on the byte slice.
+		let last_page = (self.storage.capacity() / S::ERASE_SIZE) - 1;
+
+		// `data` is the part of `bytes` contained within `page`,
+		// and `addr` in the address offset of `page` + any offset into the page as requested by `address`
+		for (data, page, addr) in (0..last_page as u32)
+			.map(move |i| Page::new(i, S::ERASE_SIZE))
+			.overlaps(bytes, offset)
+		{
+			let offset_into_page = addr.saturating_sub(page.start) as usize;
+
+			self.storage.try_read(page.start, self.merge_buffer)?;
```

Comment:
This'll read more data than needed if `merge_buffer` is bigger than one page (new just checks it's not smaller)